### ISSUE TITLE
How to refresh/reload pkcs11 library

### ISSUE THREAD

**Issue by muratkapar** [Pkcs11Interop/Pkcs11Interop#174]
Hi all and happy new year:)
I’ve been using pkcs11interop v5.0.0 for Thales nCipher HSM. Is there a way to refresh/reload/whatever the vendor pkcs11 library (cknfast.dll for ncipher) without restarting the app and with no impact to the running slots, sessions, etc.? I am asking such a solution because cknfast can only take the newly copied local key files (those are created and copied by another app) into account just after restart the app or initialize pkcs11interop from the very beginning. Any suggestion?
Thanks in advance.

**Comment by jariq** [comment]
AFAIK there are two options:

1. You can reload PKCS#11 library from within your application by calling `IPkcs11Library.Dispose` and `IPkcs11LibraryFactory.LoadPkcs11Library`. As you have correctly noted this option impacts all open sessions and turns them unusable.
2. You can search nShield user guide for `CKNFAST_ASSUME_SINGLE_PROCESS` setting. This is most likely the option you are looking for but it may decrease the performance of unmanaged PKCS#11 library. You'll find more details in the user guide.

**Comment by muratkapar** [comment]
Hi Jariq,
I’ve already tried that config parameter you specified at 2nd option, but it is unacceptable due to dramatically increase on duration of key operations.
Can an application load pkcs11library more than once in a parallel manner? If possible, one process may perform key operations while the other may try to reload itself continuously. Just an idea.
Thx,

**Comment by jariq** [comment]
> Can an application load pkcs11library more than once in a parallel manner?

I don't think so. PKCS#11 specification defines an application as a single process with single address space and one or multiple threads of control running in it. Any application becomes a "Cryptoki application" by initializing PKCS#11 library in one of its threads with a call to `C_Initialize` function. After the library has been initialized, the application can call other functions of PKCS#11 API. When the application is done using PKCS#11 API, it finalizes PKCS#11 library with a call to `C_Finalize` function and ceases to be a "Cryptoki application". From application perspective, PKCS#11 library initialization and finalization are **global events**, so it is crucial to ensure that one thread does not finalize library while other threads are still working with it. PKCS#11 function `C_Initialize` is called in constructor of class implementing `IPkcs11Library` interface and `C_Finalize` function is called when instance of such class disposed. It is crucial to ensure that two instances of this class working with same PKCS#11 library do not overlap each other.

IMO your best bet is to contact Entrust/nCipher support and ask them for help with your specific scenario or get other type of HSM that better suit your needs.

**Comment by muratkapar** [comment]
I tried a workaround and it has worked fine, thanks to https://stackoverflow.com/a/12533663. First I copied and renamed vendor specific pkcs11 library file which has been loaded through pkcs11interop. Then, as I said before, I started an async endless task in which the library files have been disposed and loaded in order, i.e. while one of them has been reloaded, the other has performed normal operations. In my case, this approach worked absolutely better than CKNFAST_ASSUME_SINGLE_PROCESS setting.
